crates/core/src/subscription/module_subscription_actor.rs

```diff
@@ -51,38 +51,43 @@ impl ModuleSubscriptions {
             queries.extend(qset);
         }
 
-        let subscription = match self.subscriptions.iter_mut().find(|s| s.queries == queries) {
+        let database_update = tokio::task::block_in_place(|| queries.eval(&self.relational_db, tx, auth))?;
+        // holds lock after `eval` so that it can be concurrent
```

Comment:
A longer comment here might be nice - I had to think about the correctness of this ordering for a few minutes, so laying it out explicitly would be helpful. We hold a read lock on the db via the tx, the module host needs a write tx to execute, etc, therefore this `queries.eval()` can't happen between a reducer being committed and it doing `broadcast_event()`. ~~I think this even holds for mvcc in the future, `begin_tx()` wouldn't lock but it would stick to the one commit it was started from; committing would put a commit on top of it but not change our tx once we've started it.~~ Oh. Wait no, this doesn't even work now, we need the secondary lock here. If `add_subscriber()` is called between commit and broadcast_event, the tx can contain the committed state, and then the broadcast_event would send a message with it and so would add_subscriber. I might push a commit to this branch with what I'm thinking, if that's ok?

Reply:
I am not sure If we can call it bug entirely, Client can receive one extra update of old subscription queuries after sending new subscription request because add_subscription has not processed yet, then receving a same data via add_subscriber must be expected.
but definitely go ahead if you can improve it.

Reply:
Or moving remove_subscription before eval will also solve it, it's just we have to acquire lock twice.

Reply:
I thought that was a bug? Wasn't that the whole reason for #792?

Reply:
> I thought that was a bug? Wasn't that the whole reason for #792?

I think main reason was other way around, i.e to receive duplicate state from broadcast after already having it from add_subscription is indeed a bug, Not really sure if vice-versa would also be a bug.

Reply:
Oh, huh. I don't know if I realized that, but that makes sense. since `add_subscriber()` is gonna return a non-incremental update anyway, it doesn't matter what gets sent before. Alright.